In [1]:
import sys
import os
sys.path.append(os.path.dirname(os.path.abspath('.')))
import networkx as nx
import numpy as np
import pandas as pd

In [2]:
import sys
import time
from simulation_functions import patients_data, nodes_position, generate_network, addBetweenPathwaysConnection, save_dataset
from simulation_nbs_functions import all_functions
from confusion_matrix import minimal_element_0_to_1, simulated_confusion_matrix
import datetime
from sklearn.model_selection import ParameterGrid
import networkx as nx
import numpy as np
import scipy.sparse as sp
import pickle
import matplotlib.pyplot as plt
import datetime
import glob
%pylab inline
%matplotlib inline

Populating the interactive namespace from numpy and matplotlib


In [3]:
pd.options.mode.chained_assignment = None  # default='warn'

In [4]:
data_folder = "input/dataset/"

# Load data & ID mapping

## Load APID PPI data (Dr. Javier DE LAS RIVAS)
Ref: __[APID interactomes: providing proteome-based interactomes with controlled quality for multiple species and derived networks](https://academic.oup.com/nar/article/44/W1/W529/2499348)__<br>

APID server: http://apid.dep.usal.es/<br>
Downloaded on Mon 2018/02(Feb)/05 <br>
Level 2: proven by 2 experiments or more:
- 147 469 interactions
- 17 547 interacting proteins


#### Raw colums:
- InteractionID	
- UniprotID_A	
- UniprotName_A	
- GeneName_A	
- UniprotID_B	
- UniprotName_B	
- GeneName_B	
- Experiments: number of experiments
- Methods: number of methods
- Publications: number of publications
- 3DStructures: number of 3D structures (PDBs)
- CurationEvents: number of curation events (including source databases)



In [6]:
df_apid = pd.read_csv(data_folder + "9606_Q2.txt", sep="\t")
print("interacting proteins number:", df_apid.shape[0])

# keep only interactions between Human-Human proteins
df_apid_HH = df_apid[df_apid.UniprotName_A.str.contains("_HUMAN") & 
                     df_apid.UniprotName_B.str.contains("_HUMAN") == True]
print("interacting Human-Human proteins number:", df_apid_HH.shape[0])

print("{} interspecific interactions removed ({}%)"
      .format(df_apid.shape[0]-df_apid_HH.shape[0],
             round((1-df_apid_HH.shape[0]/df_apid.shape[0])*100,1)))

df_apid_HH.head()

interacting proteins number: 147469
interacting Human-Human proteins number: 129827
17642 interspecific interactions removed (12.0%)


InteractionID UniprotID_A     UniprotName_A GeneName_A UniprotID_B  \
0         223552  A0A024R0Y4  A0A024R0Y4_HUMAN     TADA2L      Q8WWY3   
1         273919  A0A024R0Y4  A0A024R0Y4_HUMAN     TADA2L      O14964   
2         274865  A0A024R0Y4  A0A024R0Y4_HUMAN     TADA2L      P0DI81   
3         287424  A0A024R0Y4  A0A024R0Y4_HUMAN     TADA2L      Q6PEW1   
4         288839  A0A024R0Y4  A0A024R0Y4_HUMAN     TADA2L      Q13064   

  UniprotName_B GeneName_B  Experiments  Methods  Publications  3DStructures  \
0   PRP31_HUMAN     PRPF31            3        3             1             0   
1     HGS_HUMAN        HGS            2        2             1             0   
2   TPC2A_HUMAN    TRAPPC2            2        2             1             0   
3   ZCH12_HUMAN    ZCCHC12            2        2             1             0   
4   MKRN3_HUMAN      MKRN3            2        2             1             0   

   CurationEvents  
0               3  
1               2  
2               2  
3               2  
4               2

In [5]:
df_apid = pd.read_csv(data_folder + "9606_Q2.txt", sep="\t")
print("interacting proteins number:", df_apid.shape[0])

# keep only interactions between Human-Human proteins
df_apid_HH = df_apid[df_apid.UniprotName_A.str.contains("_HUMAN") & 
                     df_apid.UniprotName_B.str.contains("_HUMAN") == True]
print("interacting Human-Human proteins number:", df_apid_HH.shape[0])

print("{} interspecific interactions removed ({}%)"
      .format(df_apid.shape[0]-df_apid_HH.shape[0],
             round((1-df_apid_HH.shape[0]/df_apid.shape[0])*100,1)))

df_apid_HH.head()

interacting proteins number: 133993
interacting Human-Human proteins number: 115981
18012 interspecific interactions removed (13.4%)


InteractionID UniprotID_A     UniprotName_A   GeneName_A UniprotID_B  \
0         242748  A0A024R0Y4  A0A024R0Y4_HUMAN       TADA2A      Q8WWY3   
1         277739  A0A024R4Q5  A0A024R4Q5_HUMAN         TFPT      Q96D03   
2         277740  A0A024R4Q5  A0A024R4Q5_HUMAN         TFPT      P51911   
3         302485  A0A024R5S0  A0A024R5S0_HUMAN  hCG_2003792      Q9NVV9   
4         321925  A0A024R5S0  A0A024R5S0_HUMAN  hCG_2003792      Q02446   

  UniprotName_B GeneName_B  Experiments  Methods  Publications  3DStructures  \
0   PRP31_HUMAN     PRPF31            3        3             1             0   
1   DDT4L_HUMAN     DDIT4L            2        2             1             0   
2    CNN1_HUMAN       CNN1            2        2             1             0   
3   THAP1_HUMAN      THAP1            3        3             1             0   
4     SP4_HUMAN        SP4            3        3             1             0   

   CurationEvents  
0               3  
1               4  
2               4  
3               3  
4               3

In [7]:
# keep only GeneSymbol ID and UniProt ID
# then change column names
df_apid_HH = df_apid_HH[["GeneName_A", "UniprotID_A", "GeneName_B", "UniprotID_B"]]
df_apid_HH.columns = ["GeneSymbol_1", "UniProt_1", "GeneSymbol_2", "UniProt_2"]
df_apid_HH.head()

GeneSymbol_1   UniProt_1 GeneSymbol_2 UniProt_2
0       TADA2L  A0A024R0Y4       PRPF31    Q8WWY3
1       TADA2L  A0A024R0Y4          HGS    O14964
2       TADA2L  A0A024R0Y4      TRAPPC2    P0DI81
3       TADA2L  A0A024R0Y4      ZCCHC12    Q6PEW1
4       TADA2L  A0A024R0Y4        MKRN3    Q13064

In [6]:
# keep only GeneSymbol ID and UniProt ID
# then change column names
df_apid_HH = df_apid_HH[["GeneName_A", "UniprotID_A", "GeneName_B", "UniprotID_B"]]
df_apid_HH.columns = ["GeneSymbol_1", "UniProt_1", "GeneSymbol_2", "UniProt_2"]
df_apid_HH.head()

GeneSymbol_1   UniProt_1 GeneSymbol_2 UniProt_2
0       TADA2A  A0A024R0Y4       PRPF31    Q8WWY3
1         TFPT  A0A024R4Q5       DDIT4L    Q96D03
2         TFPT  A0A024R4Q5         CNN1    P51911
3  hCG_2003792  A0A024R5S0        THAP1    Q9NVV9
4  hCG_2003792  A0A024R5S0          SP4    Q02446

### There are some Gene Symbols not found

In [507]:
# APID's UniProt
apid_unip_1 = df_apid_HH.UniProt_1.drop_duplicates()
apid_unip_2 = df_apid_HH.UniProt_2.drop_duplicates()
apid_unip_rows_1 = list(apid_unip_1.index)
apid_unip_rows_2 = list(apid_unip_2.index)
apid_unip = apid_unip_1.append(apid_unip_2).drop_duplicates()

# APID's Gene Symbol
apid_geneSymb = df_apid_HH.GeneSymbol_1.append(df_apid_HH.GeneSymbol_2)
apid_geneSymb = apid_geneSymb.drop_duplicates()

print("unique UniProt number:", len(apid_unip))
print("unique Gene Symbol number:", len(apid_geneSymb))
diff = len(apid_unip)-len(apid_geneSymb)
print("there are {} fewer Gene Symbols than UniProt ID ({}%)"
      .format(diff, round((diff/len(apid_unip))*100, 1)))

unique UniProt number: 15044
unique Gene Symbol number: 14412
there are 632 fewer Gene Symbols than UniProt ID (4.2%)


In [7]:
# APID's UniProt
apid_unip_1 = df_apid_HH.UniProt_1.drop_duplicates()
apid_unip_2 = df_apid_HH.UniProt_2.drop_duplicates()
apid_unip_rows_1 = list(apid_unip_1.index)
apid_unip_rows_2 = list(apid_unip_2.index)
apid_unip = apid_unip_1.append(apid_unip_2).drop_duplicates()

# APID's Gene Symbol
apid_geneSymb = df_apid_HH.GeneSymbol_1.append(df_apid_HH.GeneSymbol_2)
apid_geneSymb = apid_geneSymb.drop_duplicates()

print("unique UniProt number:", len(apid_unip))
print("unique Gene Symbol number:", len(apid_geneSymb))
diff = len(apid_unip)-len(apid_geneSymb)
print("there are {} fewer Gene Symbols than UniProt ID ({}%)"
      .format(diff, round((diff/len(apid_unip))*100, 1)))

unique UniProt number: 14657
unique Gene Symbol number: 14241
there are 416 fewer Gene Symbols than UniProt ID (2.8%)


In [374]:
# get index of NaN values in GeneSymbol_1 & 2
null_columns = df_apid_HH.columns[df_apid_HH.isnull().any()]
print("----- NaN values in Gene Symbols -----")
print(df_apid_HH[null_columns].isnull().sum())

df_apid_HH[df_apid_HH.isnull().any(axis=1)][null_columns]
null_idx = list(df_apid_HH[df_apid_HH.isnull().any(axis=1)][null_columns].index)

df_apid_HH.loc[null_idx].head()

----- NaN values in Gene Symbols -----
GeneSymbol_1    258
GeneSymbol_2    209
dtype: int64


GeneSymbol_1   UniProt_1 GeneSymbol_2 UniProt_2
94          PIBF1  A0A087WUI6          NaN    G2XKQ0
321           NaN  A0A0B4J295         DTX2    Q86UW9
516         PROP1  A0A0G2JQ02          NaN    G2XKQ0
838           NaN  A0A1C7CYX0        PRDM7    Q9NQW5
1089         FSD2      A1L4K1          NaN    Q9H6F0

In [8]:
# get index of NaN values in GeneSymbol_1 & 2
null_columns = df_apid_HH.columns[df_apid_HH.isnull().any()]
print("----- NaN values in Gene Symbols -----")
print(df_apid_HH[null_columns].isnull().sum())

df_apid_HH[df_apid_HH.isnull().any(axis=1)][null_columns]
null_idx = list(df_apid_HH[df_apid_HH.isnull().any(axis=1)][null_columns].index)

df_apid_HH.loc[null_idx].head()

----- NaN values in Gene Symbols -----
GeneSymbol_1    126
GeneSymbol_2    135
dtype: int64


GeneSymbol_1 UniProt_1 GeneSymbol_2 UniProt_2
205         FSD2    A1L4K1          NaN    Q9H6F0
246       TARSL2    A2RTX5          NaN    B2RDX5
494          NaN    A8K2R3        KRT38    O76015
495          NaN    A8K2R3       TRIM54    Q9BYV2
496          NaN    A8K2R3       TRIM39    Q9HCM9

Sum of NaN values in GeneSymbol_1&2 (258+209=467) < 632 <br>
<b> => One Gene Symbol can have several UniProt ID </b>

## Load UniProtKB data

Complete human gene data set downloaded from __[UniProtKB](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/)__<br>
"HUMAN_9606_idmapping_selected.tab.gz" (2018/01/31)

#### Raw colums:
1. UniProtKB-AC
2. UniProtKB-ID
3. GeneID (EntrezGene)
4. RefSeq
5. GI
6. PDB
7. GO
8. UniRef100
9. UniRef90
10. UniRef50
11. UniParc
12. PIR
13. NCBI-taxon
14. MIM
15. UniGene
16. PubMed
17. EMBL
18. EMBL-CDS
19. Ensembl
20. Ensembl_TRS
21. Ensembl_PRO
22. Additional PubMed

In [9]:
# only get UniProtKB-AC and GeneID (EntrezGene)
df_unip = pd.read_csv(data_folder + "HUMAN_9606_idmapping_selected.tab", sep="\t",
                      index_col=False, usecols=[0,2], names=["UniProt_AC", "EntrezGene"])

In [10]:
df_unip.head()

UniProt_AC EntrezGene
0     P31946       7529
1     P62258       7531
2     Q04917       7533
3     P61981       7532
4     P31947       2810

## Load HGNC data
Complete HGNC dataset downloaded from __[HGNC (HUGO Gene Nomenclature Committee)](https://www.genenames.org/cgi-bin/statistics)__ <br>
"hgnc_complete_set.txt" (2018/02(Feb)/09)

#### Raw colums:
1. hgnc_id	
2. symbol	
3. name	
4. locus_group	
5. locus_type	
6. status	
7. location	
8. location_sortable	
9. alias_symbol	
10. alias_name	
11. prev_symbol	
12. prev_name	
13. gene_family	
14. gene_family_id	
15. date_approved_reserved	
16. date_symbol_changed	
17. date_name_changed	
18. date_modified	
19. entrez_id	
20. ensembl_gene_id	
21. vega_id	
22. ucsc_id	
23. ena	
24. refseq_accession	
25. ccds_id	
26. uniprot_ids	
27. pubmed_id	
28. mgd_id	
29. rgd_id	
30. lsdb	
31. cosmic	
32. omim_id	
33. mirbase	
34. homeodb	
35. snornabase	
36. bioparadigms_slc	
37. orphanet	
38. pseudogene.org	
39. horde_id	
40. merops	
41. imgt	
42. iuphar	
43. kznf_gene_catalog	
44. mamit-trnadb	
45. cd	
46. lncrnadb	
47. enzyme_id	
48. intermediate_filament_db	
49. rna_central_ids


In [11]:
# only get symbol (Gene Symbol), entrez_id (EntrezGene) and uniprot_ids
df_hgnc = pd.read_csv(data_folder + "hgnc_complete_set.txt", sep="\t",
                      index_col=False, usecols=[1,18,25])

print("----- NaN values in raw data ({} rows) -----".format(df_hgnc.symbol.size))
df_hgnc.isnull().sum()

----- NaN values in raw data (42471 rows) -----


symbol             0
entrez_id       1267
uniprot_ids    22448
dtype: int64

In [12]:
# remove row with NaN in entrez_id
df_hgnc = df_hgnc[pd.notnull(df_hgnc['entrez_id'])]

# entrez_id: float -> int
df_hgnc.entrez_id = df_hgnc.entrez_id.astype(int)
df_hgnc.head()

symbol  entrez_id uniprot_ids
0      A1BG          1      P04217
1  A1BG-AS1     503538         NaN
2      A1CF      29974      Q9NQ94
3       A2M          2      P01023
4   A2M-AS1     144571         NaN

In [13]:
print("----- NaN values after filtering ({} rows) -----".format(df_hgnc.symbol.size))
df_hgnc.isnull().sum()

----- NaN values after filtering (41204 rows) -----


symbol             0
entrez_id          0
uniprot_ids    21183
dtype: int64

## ID mapping with APID

### HGNC: from Gene Symbol to EntrezGene ID
#### Overlapping analysis between Gene Symobl and EntrezGene

In [356]:
# Gene Symbols in APID
print("all unique Gene Symbols in APID:", len(apid_geneSymb))

# Gene Symbols in HGNC 
print("all Gene Symbols in filtered HGNC:", df_hgnc.symbol.size)

all unique Gene Symbols in APID: 14412
all Gene Symbols in filtered HGNC: 42471


In [14]:
# Gene Symbols in APID
print("all unique Gene Symbols in APID:", len(apid_geneSymb))

# Gene Symbols in HGNC 
print("all Gene Symbols in filtered HGNC:", df_hgnc.symbol.size)

all unique Gene Symbols in APID: 14241
all Gene Symbols in filtered HGNC: 41204


In [357]:
intersection_geneSymb = list((set(apid_geneSymb)).intersection(set(df_hgnc.symbol)))
print("intersection of Gene Symbols:", len(intersection_geneSymb))

# print("{} of APID's UnitProt ID are not found in EntrezGene ID "
#       .format(len(apid_unip)-len(intersection_prot)))

intersection of Gene Symbols: 14125


In [17]:
intersection_geneSymb = list((set(apid_geneSymb)).intersection(set(df_hgnc.symbol)))
print("intersection of Gene Symbols:", len(intersection_geneSymb))

# print("{} of APID's UnitProt ID are not found in EntrezGene ID "
#       .format(len(apid_unip)-len(intersection_prot)))

intersection of Gene Symbols: 14082


In [650]:
# merge on APID's GeneSymbol (1 then 2) to get corresponding EntrezGene ID
# keeping the index from the 1st frame
joined1 = df_apid_HH.reset_index().merge(
    df_hgnc, how='left', left_on='GeneSymbol_1', right_on='symbol').set_index('index')
joined2 = joined1.reset_index().merge(
    df_hgnc, how='left', left_on='GeneSymbol_2', right_on='symbol').set_index('index')

# remove redundant columns
df = joined2.drop(['symbol_x', 'uniprot_ids_x', 'symbol_y', 'uniprot_ids_y'], axis=1)

# rename columns
df = df.rename(columns={'entrez_id_x': 'EntrezGene_1',
                        'entrez_id_y': 'EntrezGene_2'})
# reorder columns
cols = list(df)
cols.insert(0, cols.pop(cols.index('EntrezGene_1')))
cols.insert(3, cols.pop(cols.index('EntrezGene_2')))
df = df.loc[:, cols]

df.head()

EntrezGene_1 GeneSymbol_1   UniProt_1  EntrezGene_2 GeneSymbol_2  \
index                                                                     
0               NaN       TADA2L  A0A024R0Y4       26121.0       PRPF31   
1               NaN       TADA2L  A0A024R0Y4        9146.0          HGS   
2               NaN       TADA2L  A0A024R0Y4        6399.0      TRAPPC2   
3               NaN       TADA2L  A0A024R0Y4      170261.0      ZCCHC12   
4               NaN       TADA2L  A0A024R0Y4        7681.0        MKRN3   

      UniProt_2  
index            
0        Q8WWY3  
1        O14964  
2        P0DI81  
3        Q6PEW1  
4        Q13064

In [18]:
# merge on APID's GeneSymbol (1 then 2) to get corresponding EntrezGene ID
# keeping the index from the 1st frame
joined1 = df_apid_HH.reset_index().merge(
    df_hgnc, how='left', left_on='GeneSymbol_1', right_on='symbol').set_index('index')
joined2 = joined1.reset_index().merge(
    df_hgnc, how='left', left_on='GeneSymbol_2', right_on='symbol').set_index('index')

# remove redundant columns
df = joined2.drop(['symbol_x', 'uniprot_ids_x', 'symbol_y', 'uniprot_ids_y'], axis=1)

# rename columns
df = df.rename(columns={'entrez_id_x': 'EntrezGene_1',
                        'entrez_id_y': 'EntrezGene_2'})
# reorder columns
cols = list(df)
cols.insert(0, cols.pop(cols.index('EntrezGene_1')))
cols.insert(3, cols.pop(cols.index('EntrezGene_2')))
df = df.loc[:, cols]

df.head()

EntrezGene_1 GeneSymbol_1   UniProt_1  EntrezGene_2 GeneSymbol_2  \
index                                                                     
0            6871.0       TADA2A  A0A024R0Y4       26121.0       PRPF31   
1           29844.0         TFPT  A0A024R4Q5      115265.0       DDIT4L   
2           29844.0         TFPT  A0A024R4Q5        1264.0         CNN1   
3               NaN  hCG_2003792  A0A024R5S0       55145.0        THAP1   
4               NaN  hCG_2003792  A0A024R5S0        6671.0          SP4   

      UniProt_2  
index            
0        Q8WWY3  
1        Q96D03  
2        P51911  
3        Q9NVV9  
4        Q02446

In [651]:
df.isnull().sum().GeneSymbol_1

258

In [19]:
df.isnull().sum().GeneSymbol_1

126

In [652]:
# get index of NaN values in EntrezGene_1 & 2
# entrez_col = df.columns[df.columns.str.contains("EntrezGene")]
# df[df.isnull().any(axis=1)][entrez_col]
# nan_idx = list(df[df.isnull().any(axis=1)][entrez_col].index)
# df.loc[nan_idx]

In [20]:
# Get Gene Symbols with which EntrezGene are NaN
df_geneSymb_entrezNan_1 = df[df.EntrezGene_1.isnull()].GeneSymbol_1
df_geneSymb_entrezNan_2 = df[df.EntrezGene_2.isnull()].GeneSymbol_2

# remove duplicates
df_geneSymb_entrezNan_1 = df_geneSymb_entrezNan_1.drop_duplicates()
df_geneSymb_entrezNan_2 = df_geneSymb_entrezNan_2.drop_duplicates()

df_geneSymb_entrezNan = df_geneSymb_entrezNan_1.append(df_geneSymb_entrezNan_2).drop_duplicates()

print("There are {} ({}%) missing EntrezGene ID via Gene Symbol ID mapping"
     .format(len(df_geneSymb_entrezNan),
             round((len(df_geneSymb_entrezNan)/len(apid_geneSymb))*100, 1)))

There are 159 (1.1%) missing EntrezGene ID via Gene Symbol ID mapping


### UniProtKB-AC: from UniProt ID to EntrezGene ID
#### For Gene Symbols not previously found in APID

In [21]:
# UniProtKB
unip_entrez = df_unip.UniProt_AC.drop_duplicates()
print('all UniProtKB ID:', len(unip_entrez))

# ADIP
apid_unip = df_apid_HH.UniProt_1.append(df_apid_HH.UniProt_2)
apid_unip = apid_unip.drop_duplicates()

overlap_uniprot_adip = (len(apid_unip)/len(unip_entrez))*100
print("{} interacting proteins in ADIP ({}%)"
      .format(len(apid_unip), round(overlap_uniprot_adip, 1)))

all UniProtKB ID: 161566
14657 interacting proteins in ADIP (9.1%)


In [22]:
# data with NaN in GeneSymbol_1 or 2
df_apid_geneSymb_nan1 = df[df.GeneSymbol_1.isnull()]
df_apid_geneSymb_nan2 = df[df.GeneSymbol_2.isnull()]

# merge on APID's UniProt_1 and 2 SEPARATELY to get corresponding EntrezGene ID
# keeping the index from the 1st frame
joined_1 = df_apid_geneSymb_nan1.reset_index().merge(
    df_unip, how='left', left_on='UniProt_1', right_on='UniProt_AC').set_index('index')
joined_2 = df_apid_geneSymb_nan2.reset_index().merge(
    df_unip, how='left', left_on='UniProt_2', right_on='UniProt_AC').set_index('index')

# replace new EntrezGene values found in UniProtKB
df.EntrezGene_1[joined_1.index] = joined_1.EntrezGene
df.EntrezGene_2[joined_2.index] = joined_2.EntrezGene

In [23]:
# rows with both GeneSymbol_1&2 are NaN
df.loc[joined_1.index.intersection(joined_2.index)]

EntrezGene_1 GeneSymbol_1 UniProt_1 EntrezGene_2 GeneSymbol_2 UniProt_2
index                                                                        
598            NaN          NaN    B2R8Y4          NaN          NaN    G2XKQ0
68812          NaN          NaN    Q59GP6          NaN          NaN    G2XKQ0

In [657]:
nan_before = df.isnull().sum().GeneSymbol_1 + df.isnull().sum().GeneSymbol_2 - 1
nan_after = joined_1.EntrezGene.isnull().sum() + joined_2.EntrezGene.isnull().sum() - 1
print("NaN Entrez Gene before mapping with UniProtKB:", nan_before)
print("NaN Entrez Gene after mapping with UniProtKB:", nan_after)
print("newly found: {} ({}%)"
      .format(nan_before-nan_after, 100-round(nan_after/nan_before, 1)*100))

NaN Entrez Gene before mapping with UniProtKB: 466
NaN Entrez Gene after mapping with UniProtKB: 379
newly found: 87 (20.0%)


In [25]:
nan_before = df.isnull().sum().GeneSymbol_1 + df.isnull().sum().GeneSymbol_2
nan_after = joined_1.EntrezGene.isnull().sum() + joined_2.EntrezGene.isnull().sum() 
print("NaN Entrez Gene before mapping with UniProtKB:", nan_before)
print("NaN Entrez Gene after mapping with UniProtKB:", nan_after)
print("newly found: {} ({}%)"
      .format(nan_before-nan_after, 100-round(nan_after/nan_before, 1)*100))

NaN Entrez Gene before mapping with UniProtKB: 261
NaN Entrez Gene after mapping with UniProtKB: 175
newly found: 86 (30.0%)


## Filtering

### Remove EntrezGene NaN rows

In [717]:
df_entrez = df[df['EntrezGene_1'].notnull()]
df_entrez = df_entrez[df_entrez['EntrezGene_2'].notnull()]
df_entrez.EntrezGene_1 = df_entrez.EntrezGene_1.astype(int)
df_entrez.EntrezGene_2 = df_entrez.EntrezGene_2.astype(int)

print("removed PPI number: {} ({}%)".format(df.shape[0]-df_entrez.shape[0],
                                          round(100-df_entrez.shape[0]/df.shape[0]*100,1)))

# unique number
df_entrez_unique_unip = df_entrez.UniProt_1.append(df_entrez.UniProt_2)
df_entrez_unique_unip = df_entrez_unique_unip.drop_duplicates()

df_entrez_unique_entrez = df_entrez.EntrezGene_1.append(df_entrez.EntrezGene_2)
df_entrez_unique_entrez = df_entrez_unique_entrez.drop_duplicates()

df_entrez_unique_geneSymb = df_entrez.GeneSymbol_1.append(df_entrez.GeneSymbol_2)
df_entrez_unique_geneSymb = df_entrez_unique_geneSymb.drop_duplicates()

print("unique UniProt:", len(df_entrez_unique_unip))
print("unique Entrez Gene:", len(df_entrez_unique_entrez))
print("unique Gene Symbol:", len(df_entrez_unique_geneSymb))

removed PPI number: 2647 (2.0%)
unique UniProt: 14652
unique Entrez Gene: 14117
unique Gene Symbol: 14113


In [26]:
df_entrez = df[df['EntrezGene_1'].notnull()]
df_entrez = df_entrez[df_entrez['EntrezGene_2'].notnull()]
df_entrez.EntrezGene_1 = df_entrez.EntrezGene_1.astype(int)
df_entrez.EntrezGene_2 = df_entrez.EntrezGene_2.astype(int)

print("removed PPI number: {} ({}%)".format(df.shape[0]-df_entrez.shape[0],
                                          round(100-df_entrez.shape[0]/df.shape[0]*100,1)))

# unique number
df_entrez_unique_unip = df_entrez.UniProt_1.append(df_entrez.UniProt_2)
df_entrez_unique_unip = df_entrez_unique_unip.drop_duplicates()

df_entrez_unique_entrez = df_entrez.EntrezGene_1.append(df_entrez.EntrezGene_2)
df_entrez_unique_entrez = df_entrez_unique_entrez.drop_duplicates()

df_entrez_unique_geneSymb = df_entrez.GeneSymbol_1.append(df_entrez.GeneSymbol_2)
df_entrez_unique_geneSymb = df_entrez_unique_geneSymb.drop_duplicates()

print("unique UniProt:", len(df_entrez_unique_unip))
print("unique Entrez Gene:", len(df_entrez_unique_entrez))
print("unique Gene Symbol:", len(df_entrez_unique_geneSymb))

removed PPI number: 1067 (0.9%)
unique UniProt: 14426
unique Entrez Gene: 14080
unique Gene Symbol: 14076


In [28]:
# save
filename = ('input/dataset/PPI_APID_{}.csv'.
            format(datetime.datetime.now().strftime("%Y_%m_%d")))
df_entrez.to_csv(filename, sep='\t', index=False)

# PPI network adjacency matrix
around 30-40 secondes to generate a new adjacency matrix

In [63]:
import glob
from scipy.io import loadmat, savemat

In [75]:
def load_PPI(data_folder, ppi_data, load_gene_id_ppi=True):
    print(' ==== load_PPI ')
    filename = 'PPI_' + ppi_data + '.mat'
    loadfile = loadmat(data_folder + filename)
    network = loadfile['adj_mat'].astype(np.float32)
    if load_gene_id_ppi:
        print(' ==== load_gene_id_ppi ')
        gene_id_ppi = (loadfile['entrez_id'].flatten()).tolist()
        return gene_id_ppi, network
    else:
        return network

In [55]:
def coordinate(prot_list, all_list):
    coo_list = []
    for prot in prot_list:
        i = all_list.index(prot)
        coo_list.append(i)
    return coo_list

In [69]:
def create_adjacency_matrix(prot1, prot2):
    # prot1 and prot2 are two columns including Entrez gene ID, from PPI data
    edge_list = np.vstack((prot1, prot2)).T
    gene_id_ppi = (edge_list.flatten()).tolist()
    gene_id_ppi = list(set(gene_id_ppi))

    # From ID list to coordinate list
    print(' ==== coordinates ')
    coo1 = coordinate(prot1.tolist(), gene_id_ppi)
    coo2 = coordinate(prot2.tolist(), gene_id_ppi)

    # Adjacency matrix
    print(' ==== Adjacency matrix ')
    n = len(gene_id_ppi)
    weight = np.ones(len(coo1))  # if interaction -> 1
    network = sp.coo_matrix((weight, (coo1, coo2)), shape=(n, n)) 
    network = network + network.T  # symmetric matrix
    savemat(PPI_file, {'adj_mat': network, 'entrez_id': gene_id_ppi},
            do_compression=True)
    return gene_id_ppi, network

In [70]:
def load_PPI_Y2H_or_APID(data_folder, ppi_data):
    print(' ==== load_PPI ')
    PPI_file = data_folder + 'PPI_' + ppi_data + '.mat'
    existance_file = os.path.exists(PPI_file)

    if existance_file:
        print('***** PPI_{} file already exists *****'.format(ppi_data))
        gene_id_ppi, network = load_PPI(
            data_folder, ppi_data, load_gene_id_ppi=True)

    else:
        print('PPI_{} file is calculating.....'.format(ppi_data))
        if ppi_data == "Y2H":
            data = genfromtxt(data_folder+'PPI_Y2H_raw.tsv',
                          delimiter='\t', dtype=int)
            # List of all proteins with Entrez gene ID
            prot1 = data[1:, 0]
            prot2 = data[1:, 1]
        else:
            newest_file = max(glob.iglob(data_folder + 'PPI_APID*.csv'),
                              key=os.path.getctime)
            data = pd.read_csv(newest_file, sep="\t")
            prot1 = data.EntrezGene_1
            prot2 = data.EntrezGene_2
        
        gene_id_ppi, network = create_adjacency_matrix(prot1, prot2)

    return gene_id_ppi, network

In [71]:
%time gene_id_ppi, network = load_PPI_Y2H_or_APID(data_folder, ppi_data="APID")

 ==== load_PPI 
PPI_APID file is calculating.....
 ==== coordinates 
 ==== Adjacency matrix 
CPU times: user 39.8 s, sys: 254 ms, total: 40 s
Wall time: 42.1 s


In [76]:
%time gene_id_ppi, network = load_PPI_Y2H_or_APID(data_folder, ppi_data="APID")

 ==== load_PPI 
***** PPI_APID file already exists *****
 ==== load_PPI 
 ==== load_gene_id_ppi 
CPU times: user 32.9 ms, sys: 5.28 ms, total: 38.2 ms
Wall time: 37.3 ms
